### ISSUE TITLE
Tycho 3.0.0 fails to build - java.lang.NoSuchMethodError: void org.eclipse.equinox.internal.p2.repository.helpers.ChecksumProducer

### ISSUE THREAD

**Issue by Phillipus** [eclipse-tycho/tycho#1814]
I use Tycho 3.0.0 to build our RCP app. Today it no longer works resulting in this error:

[ERROR] Failed to execute goal org.eclipse.tycho:tycho-p2-repository-plugin:3.0.0:assemble-repository (default-assemble-repository) on project com.archimatetool.editor.product: Execution default-assemble-repository of goal org.eclipse.tycho:tycho-p2-repository-plugin:3.0.0:assemble-repository failed: An API incompatibility was encountered while executing org.eclipse.tycho:tycho-p2-repository-plugin:3.0.0:assemble-repository: java.lang.NoSuchMethodError: 'void org.eclipse.equinox.internal.p2.repository.helpers.ChecksumProducer.<init>(java.lang.String, java.lang.String, java.lang.String)'

```
Caused by: java.lang.NoSuchMethodError: 'void org.eclipse.equinox.internal.p2.repository.helpers.ChecksumProducer.<init>(java.lang.String, java.lang.String, java.lang.String)'
    at org.eclipse.equinox.internal.p2.artifact.processors.checksum.ChecksumUtilities.calculateChecksums (ChecksumUtilities.java:124)
    at org.eclipse.equinox.p2.internal.repository.tools.RecreateRepositoryApplication.recreateRepository (RecreateRepositoryApplication.java:128)
    at org.eclipse.equinox.p2.internal.repository.tools.RecreateRepositoryApplication.run (RecreateRepositoryApplication.java:48)
    at org.eclipse.tycho.p2.tools.mirroring.MirrorApplicationServiceImpl.recreateArtifactRepository (MirrorApplicationServiceImpl.java:223)
    at org.eclipse.tycho.p2.tools.mirroring.MirrorApplicationServiceImpl.mirrorReactor (MirrorApplicationServiceImpl.java:183)
    at org.eclipse.tycho.plugins.p2.repository.AssembleRepositoryMojo.execute (AssembleRepositoryMojo.java:217)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

I haven't changed anything. ~It works OK on Windows and Mac Intel.~

Here is our main pom:

https://github.com/archimatetool/archi/blob/master/pom.xml

And the one for the *.product file:

https://github.com/archimatetool/archi/blob/master/com.archimatetool.editor.product/pom.xml

Using Tycho 3.0.0

I have no idea what could be causing this. I've deleted the .m2 directory and still the same result.

Did something change upstream?

**Comment by Phillipus** [comment]
Tested with Tycho 2.7.5 and that is working OK.

**Comment by Phillipus** [comment]
There are two versions of `org.eclipse.equinox.p2.repository` in my .m2 directory:

 `.m2/repository/p2/osgi/bundle/org.eclipse.equinox.p2.repository/2.6.200.v20220819-1949/org.eclipse.equinox.p2.repository-2.6.200.v20220819-1949.jar`

and

`.m2/repository/p2/osgi/bundle/org.eclipse.equinox.p2.repository/2.6.300.v20221030-1923/org.eclipse.equinox.p2.repository-2.6.300.v20221030-1923.jar`

The later one has the constructor `public ChecksumProducer(String id, String algorithm, String providerName)`, but the earlier one does not. So it seems to be picking the wrong one, but only on Mac Silicon.

**Comment by Phillipus** [comment]
I checked an older .m2 directory from a backup a few days ago and it only has the `2.6.200.v20220819-1949` version. So something has changed upstream recently to add a later version `2.6.300.v20221030-1923` and for some reason this is not working on Mac Silicon.

**Comment by Phillipus** [comment]
Edit: it's not just Mac Silicon. It's now doing the same on my Windows machine.

**Comment by rgrunber** [comment]
I'm able to reproduce locally with JDT-LS. I had to use `-Dmaven.repo.local=$(pwd)/.m2` to get a completely fresh local cache.

**Comment by cdietrich** [comment]
We see the same 
https://github.com/itemis/xtext-reference-projects/actions/runs/3635745670/jobs/6135114598

**Comment by kwin** [comment]
Root issue #1464

**Comment by laeubi** [comment]
> Root issue #1464

Nope it is (and hopefully getting better when we upgrade to the 2022-12 release), Tycho is not the only consumer affected by this
- https://github.com/eclipse-platform/eclipse.platform.releng/issues/119

The only fix I know of is to limit the version of the effending dependency in the `pom.xml` when defining the plugin but maybe @olamy or @kwin know about a better alternative?

**Comment by laeubi** [comment]
> There are two versions of `org.eclipse.equinox.p2.repository` in my .m2 directory:

Can you try out the following (might differ depending on what tycho plugin you exactly use):

```
<plugin>
	<groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-p2-repository-plugin</artifactId>
        <version>3.0.0</version>

	<dependency>
	  <groupId>org.eclipse.platform</groupId>
	  <artifactId>org.eclipse.equinox.p2.core</artifactId>
	  <version>2.6.200</version>
	</dependency>
</plugin>
```

**Comment by laeubi** [comment]
I'll prepare a bugfix 3.0.1 release so please help in collecting / testing see:
https://github.com/eclipse-tycho/tycho/discussions/1401#discussioncomment-4329675

**Comment by hd42** [comment]
Using
```
                    <dependencies>
                        <dependency>
                            <groupId>org.eclipse.platform</groupId>
                            <artifactId>org.eclipse.equinox.p2.core</artifactId>
                            <version>2.6.200</version>
                        </dependency>
                    </dependencies>
```
(Note the `<dependencies>` tag missing from your example, @laeubi) does not work for our build (I tried it in the `<pluginManagement>` as well as in the `<plugins>` sections). Any progress on the SNAPSHOT?

**Comment by laeubi** [comment]
Yes... that would have been too easy :-\
I'm currently prepare a complete upgrade to 2022-12 for the tycho-3.0.x snapshot now, this seems working much better that try to fix the old dependencies.

**Comment by joakime** [comment]
> > There are two versions of `org.eclipse.equinox.p2.repository` in my .m2 directory:
> 
> Can you try out the following (might differ depending on what tycho plugin you exactly use):
> 
> ```
> <plugin>
> 	<groupId>org.eclipse.tycho</groupId>
>         <artifactId>tycho-p2-repository-plugin</artifactId>
>         <version>3.0.0</version>
> 
> 	<dependency>
> 	  <groupId>org.eclipse.platform</groupId>
> 	  <artifactId>org.eclipse.equinox.p2.core</artifactId>
> 	  <version>2.6.200</version>
> 	</dependency>
> </plugin>
> ```

This combination produces the following error/warning ...

```
[INFO] --- tycho-p2-repository-plugin:3.0.0:assemble-maven-repository (maven-p2-site) @ jetty-p2 ---
[ERROR] [93b6413b-c6ef-4bb1-b5ad-f5f04ab47384][plugin>org.eclipse.tycho:tycho-p2-repository-plugin:3.0.0] org.eclipse.tycho.p2.tools.impl
org.osgi.framework.BundleException: Could not resolve module: org.eclipse.tycho.p2.tools.impl [41]
  Unresolved requirement: Require-Bundle: org.eclipse.equinox.p2.director.app
    -> Bundle-SymbolicName: org.eclipse.equinox.p2.director.app; bundle-version="1.2.200.v20220802-0626"; tags:List<String>="osgi.connect"; singleton:="true"
       org.eclipse.equinox.p2.director.app [53]
         Unresolved requirement: Import-Package: org.eclipse.equinox.p2.core; version="[2.8.0,3.0.0)"
```

**Comment by joakime** [comment]
Eclipse Jetty hit this as well.

Our changed plugin definition to prevent the plugin from failing the build (don't know if it's working properly though) ...

``` xml
      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-p2-repository-plugin</artifactId>
        <version>3.0.0</version>
        <dependencies>
          <dependency>
            <groupId>org.eclipse.platform</groupId>
            <artifactId>org.eclipse.equinox.p2.core</artifactId>
            <version>2.9.200</version>
          </dependency>
        </dependencies>
```

**Comment by Phillipus** [comment]
> Eclipse Jetty hit this as well.
> 
> Our changed plugin definition to prevent the plugin from failing the build (don't know if it's working properly though) ...
> 
> ```
>       <plugin>
>         <groupId>org.eclipse.tycho</groupId>
>         <artifactId>tycho-p2-repository-plugin</artifactId>
>         <version>3.0.0</version>
>         <dependencies>
>           <dependency>
>             <groupId>org.eclipse.platform</groupId>
>             <artifactId>org.eclipse.equinox.p2.core</artifactId>
>             <version>2.9.200</version>
>           </dependency>
>         </dependencies>
> ```

Even though I only see version `2.6.200.v20220819-1949` in `.m2` I'm still getting the same error.

**Comment by laeubi** [comment]
> Any progress on the SNAPSHOT?

I hope to sorted out all stuff now and the build is running:
https://ci.eclipse.org/tycho/job/tycho-github/job/tycho-3.0.x/23/

if that is successful the [new 3.0.1.-SNAPSHOT](https://github.com/eclipse/tycho/wiki#getting-tycho-snapshots) should be available.

**Comment by Phillipus** [comment]
I tested the new 3.0.1-SNAPSHOT and it's working now. Thanks!

**Comment by sewe** [comment]
FWIW, 3.0.1-SNAPSHOT works for our (moderately complex) builds, too.

**Comment by joakime** [comment]
> > Eclipse Jetty hit this as well.
> > Our changed plugin definition to prevent the plugin from failing the build (don't know if it's working properly though) ...
> > ```
> >       <plugin>
> >         <groupId>org.eclipse.tycho</groupId>
> >         <artifactId>tycho-p2-repository-plugin</artifactId>
> >         <version>3.0.0</version>
> >         <dependencies>
> >           <dependency>
> >             <groupId>org.eclipse.platform</groupId>
> >             <artifactId>org.eclipse.equinox.p2.core</artifactId>
> >             <version>2.9.200</version>
> >           </dependency>
> >         </dependencies>
> > ```
> 
> Even though I only see version `2.6.200.v20220819-1949` in `.m2` I'm still getting the same error.

2.9.200 is available on Maven Central.

* Official Search Results - https://search.maven.org/artifact/org.eclipse.platform/org.eclipse.equinox.p2.core
* Browsing of main repository - https://repo1.maven.org/maven2/org/eclipse/platform/org.eclipse.equinox.p2.core/2.9.200/

**Comment by laeubi** [comment]
Thanks all for verification, the snapshot is now released and should shortly show up in central.

> upport to 3.1.0-SNAPSHOT will follow i guess

Currently there is no plan for Tycho 3.1 release

**Comment by cdietrich** [comment]
@laeubi so we also should stop testing 3.1.0-SNAPSHOT?

**Comment by Phillipus** [comment]
Closing this as fixed in 3.0.1. Thanks for the fast turnaround!

**Comment by olamy** [comment]
Thanks for the quick fix!
